src/stdatamodels/model_base.py

```diff
@@ -585,10 +585,15 @@ def to_asdf(self, init, *args, **kwargs):
             `~asdf.AsdfFile.write_to`.
         """
         self.on_save(init)
-        self.validate()
-        asdffile = self.open_asdf(self._instance, **kwargs)
+        self.validate()  # required to trigger ValidationWarning
+        tree = convert_fitsrec_to_array_in_tree(self._instance)
+        # don't open_asdf(tree) as this will cause a second validation of the tree
+        # instead open an empty tree, then assign to the hidden '_tree'
+        asdffile = self.open_asdf(None, **kwargs)
+        asdffile._tree = tree
```

Comment:
The assignment to `af._tree` instead of `af.tree` is for the same reasons as described in the comment and in https://github.com/spacetelescope/stdatamodels/pull/205/files#r1316260255